File: apps/windows/TeslaSync.App/feature-views/DrivingTips/DrivingTips.Model.cs
using System.Globalization;
using TeslaSync.App.Core;
using TeslaSync.App.Core.Notifications;

namespace TeslaSync.App.FeatureViews;

/// <summary>
/// The mutually-exclusive render branch of the <c>DrivingTips</c> surface — the native union of the states the
/// P2 feature-view contract requires for the driving-dynamics recommendations panel
/// (web/src/features/driving/components/driving-dynamics/DrivingTips.tsx). The web component is a pure
/// presentational child: it takes already-resolved <c>motorStats</c> / <c>throttleStyle</c> props and performs
/// no fetching, so the parent Driving-Dynamics page owns the query lifecycle and supplies the active state. The
/// native surface reproduces the full loading / ready / empty / error / stale / offline matrix the prompt
/// mandates; every member maps onto a visible surface (the glass recommendations panel, a skeleton, or a retry
/// affordance) and none is ever hidden behind a <c>{data &amp;&amp; …}</c> guard. <see cref="Empty"/> mirrors the
/// web rendering its <c>tipNoData</c> branch when <c>motorStats</c> is null (the "Drive your vehicle to start
/// collecting dynamics data." row), never a blank box.
/// </summary>
public enum DrivingTipsState
{
    /// <summary>The parent query is in flight and no motor stats have arrived yet — skeleton chrome.</summary>
    Loading,

    /// <summary>Resolved motor stats to render — the lightbulb panel over the derived tip rows (web fall-through).</summary>
    Ready,

    /// <summary>Resolved with no motor stats — the panel over the single "drive to collect data" row, never blank.</summary>
    Empty,

    /// <summary>The query failed with no usable snapshot — a retriable error surface.</summary>
    Error,

    /// <summary>Showing a snapshot older than the freshness window — the tips plus a stale chip.</summary>
    Stale,

    /// <summary>No connectivity — the last cached tips plus an offline chip.</summary>
    Offline,
}

/// <summary>
/// One driving-style recommendation the <c>DrivingTips</c> surface can show — the native, localizer-free
/// enumeration of the exact <c>t('dynamics.tip…')</c> branches the web <c>useMemo</c> derives
/// (web/src/features/driving/components/driving-dynamics/DrivingTips.tsx). Keeping the <em>selection</em> as an
/// enum (separate from its localized text) lets the threshold logic be asserted headlessly, byte-for-byte
/// against the web, without a resource host.
/// </summary>
public enum DrivingTipKind
{
    /// <summary>web <c>dynamics.tipNoData</c> — shown when there are no motor stats yet.</summary>
    NoData,

    /// <summary>web <c>dynamics.tipEaseAccel</c> — high average power, ease into the accelerator.</summary>
    EaseAccel,

    /// <summary>web <c>dynamics.tipBrakeEarly</c> — high average power, brake earlier for regen.</summary>
    BrakeEarly,

    /// <summary>web <c>dynamics.tipSmoothThrottle</c> — moderate average power, smooth transitions.</summary>
    SmoothThrottle,

    /// <summary>web <c>dynamics.tipCoast</c> — moderate average power, lift off to coast on regen.</summary>
    Coast,

    /// <summary>web <c>dynamics.tipGreat</c> — low average power, excellent style.</summary>
    Great,

    /// <summary>web <c>dynamics.tipKeep</c> — low average power, keep it consistent.</summary>
    Keep,

    /// <summary>web <c>dynamics.tipThermal</c> — appended when the peak motor temperature runs high.</summary>
    Thermal,
}

/// <summary>
/// Derives the ordered driving-style recommendations from a drive's motor statistics — the 1:1 native port of the
/// web <c>DrivingTips</c> <c>useMemo</c> body (web/src/features/driving/components/driving-dynamics/DrivingTips.tsx)
/// and the surface's "cached telemetry → projection" data adapter. The branch thresholds are reproduced exactly:
/// with no stats the single <see cref="DrivingTipKind.NoData"/> row is returned; otherwise the average motor
/// power (kW) selects the ease/brake (&gt; <see cref="EaseThresholdKw"/>), smooth/coast
/// (&gt; <see cref="SmoothThresholdKw"/>) or great/keep pair, and a peak motor temperature above
/// <see cref="ThermalTipCeilingCelsius"/> appends the thermal caution. All comparisons are strict <c>&gt;</c>
/// exactly like the web. Power is the SI-derived kilowatt figure and temperature the raw SI Celsius the API
/// delivers, so the recommendations never change with the user's display units. Pure logic — unit-tested without
/// a UI host.
/// </summary>
public static class DrivingTipsAdapter
{
    /// <summary>Average motor power (kW) above which the high-power ease/brake tips show (web <c>avgPower &gt; 80</c>).</summary>
    public const double EaseThresholdKw = 80;

    /// <summary>Average motor power (kW) above which the moderate smooth/coast tips show (web <c>avgPower &gt; 20</c>).</summary>
    public const double SmoothThresholdKw = 20;

    /// <summary>Peak motor temperature (°C) above which the thermal caution is appended (web <c>maxMotorTemp &gt; 120</c>).</summary>
    public const double ThermalTipCeilingCelsius = 120;

    /// <summary>
    /// Select the ordered recommendations for <paramref name="stats"/> — the web <c>useMemo</c> list. A null
    /// <paramref name="stats"/> yields the single <see cref="DrivingTipKind.NoData"/> row.
    /// </summary>
    /// <param name="stats">The aggregated motor stats (web <c>motorStats</c>), or null.</param>
    /// <returns>The ordered recommendation kinds, never empty.</returns>
    public static IReadOnlyList<DrivingTipKind> Select(MotorEfficiencyStats? stats)
    {
        var tips = new List<DrivingTipKind>(3);

        if (stats is not { } s)
        {
            tips.Add(DrivingTipKind.NoData);
            return tips;
        }

        if (s.AvgPowerKw > EaseThresholdKw)
        {
            tips.Add(DrivingTipKind.EaseAccel);
            tips.Add(DrivingTipKind.BrakeEarly);
        }
        else if (s.AvgPowerKw > SmoothThresholdKw)
        {
            tips.Add(DrivingTipKind.SmoothThrottle);
            tips.Add(DrivingTipKind.Coast);
        }
        else
        {
            tips.Add(DrivingTipKind.Great);
            tips.Add(DrivingTipKind.Keep);
        }

        if (s.MaxMotorTempCelsius > ThermalTipCeilingCelsius)
        {
            tips.Add(DrivingTipKind.Thermal);
        }

        return tips;
    }

    /// <summary>
    /// The i18n key and English fallback for a recommendation — the exact <c>t(key, fallback)</c> pair the web
    /// component passes for each tip. The fallback is rendered verbatim when the key is absent from the catalog.
    /// </summary>
    /// <param name="kind">The recommendation to resolve.</param>
    /// <returns>The i18n key and its English fallback.</returns>
    public static (string Key, string Fallback) Resource(DrivingTipKind kind) => kind switch
    {
        DrivingTipKind.NoData => (
            "dynamics.tipNoData",
            "Drive your vehicle to start collecting dynamics data."),
        DrivingTipKind.EaseAccel => (
            "dynamics.tipEaseAccel",
            "Ease into the accelerator \u2014 gradual inputs save energy and tire wear."),
        DrivingTipKind.BrakeEarly => (
            "dynamics.tipBrakeEarly",
            "Brake earlier and lighter to improve regen capture."),
        DrivingTipKind.SmoothThrottle => (
            "dynamics.tipSmoothThrottle",
            "Smooth throttle transitions can improve efficiency by 10\u201315%."),
        DrivingTipKind.Coast => (
            "dynamics.tipCoast",
            "Lift off the pedal earlier to let regen do the work."),
        DrivingTipKind.Great => (
            "dynamics.tipGreat",
            "Excellent driving style! Maintaining this maximizes range and comfort."),
        DrivingTipKind.Keep => (
            "dynamics.tipKeep",
            "Keep monitoring your scores \u2014 consistency is key."),
        _ => (
            "dynamics.tipThermal",
            "Motor temps are running high \u2014 consider easing off sustained high power."),
    };

    /// <summary>Resolve a recommendation to its localized text through the i18n facade.</summary>
    /// <param name="kind">The recommendation to localize.</param>
    /// <param name="localizer">The i18n facade.</param>
    /// <returns>The localized recommendation text.</returns>
    public static string Localize(DrivingTipKind kind, ILocalizer localizer)
    {
        ArgumentNullException.ThrowIfNull(localizer);
        (string key, string fallback) = Resource(kind);
        return localizer.GetString(key, fallback);
    }
}

/// <summary>
/// The render-time data model the <c>DrivingTips</c> view binds to — the native analogue of the web component's
/// props (<c>motorStats</c> / <c>throttleStyle</c>,
/// web/src/features/driving/components/driving-dynamics/DrivingTips.tsx) plus the parent-supplied lifecycle
/// <see cref="Status"/> and freshness flags. The view never performs HTTP; the parent Driving-Dynamics state
/// holder fills this in (the native P1/S8 seam). Motor stats stay SI; the recommendations and the row icon are
/// derived at projection time. <see cref="Style"/> is optional — when the stats are present and no style is
/// supplied, the projection derives it from the average power exactly like the web parent
/// (<c>throttleStyle = motorStats ? getThrottleStyle(motorStats.avgPower) : null</c>). Pure data so the
/// projection is unit-tested without a UI host.
/// </summary>
/// <param name="Status">The parent-supplied lifecycle state.</param>
/// <param name="Stats">The aggregated motor stats (web <c>motorStats</c>), or null when there are none.</param>
/// <param name="Style">The driving style (web <c>throttleStyle</c>), or null to derive from the average power.</param>
/// <param name="UpdatedAt">Last successful update timestamp surfaced in the freshness chip.</param>
/// <param name="IsFetching">True while a background refresh is in flight.</param>
/// <param name="ErrorMessage">Already-localized error message for the error / offline surfaces, when set.</param>
public sealed record DrivingTipsModel(
    DrivingTipsState Status,
    MotorEfficiencyStats? Stats,
    ThrottleStyle? Style,
    DateTimeOffset? UpdatedAt = null,
    bool IsFetching = false,
    string? ErrorMessage = null)
{
    /// <summary>The initial model: the parent query is in flight and no stats have arrived yet.</summary>
    public static DrivingTipsModel Loading() =>
        new(DrivingTipsState.Loading, null, null);

    /// <summary>A resolved model with no motor stats — the panel over the "drive to collect data" row.</summary>
    public static DrivingTipsModel Empty() =>
        new(DrivingTipsState.Empty, null, null);

    /// <summary>A hard-failure model (no usable snapshot) carrying an optional already-localized message.</summary>
    /// <param name="message">An already-localized error message, or null for the default copy.</param>
    public static DrivingTipsModel Failed(string? message = null) =>
        new(DrivingTipsState.Error, null, null, ErrorMessage: message);

    /// <summary>A fresh resolved model carrying the motor stats the recommendations derive from.</summary>
    /// <param name="stats">The aggregated motor stats.</param>
    /// <param name="style">The driving style, or null to derive from the average power.</param>
    /// <param name="updatedAt">The freshness timestamp.</param>
    /// <param name="isFetching">True while a background refresh is in flight.</param>
    public static DrivingTipsModel Ready(
        MotorEfficiencyStats stats,
        ThrottleStyle? style = null,
        DateTimeOffset? updatedAt = null,
        bool isFetching = false)
    {
        ArgumentNullException.ThrowIfNull(stats);
        return new(DrivingTipsState.Ready, stats, style, updatedAt, isFetching);
    }

    /// <summary>A stale snapshot (older than the freshness window) carrying the cached motor stats.</summary>
    /// <param name="stats">The cached motor stats.</param>
    /// <param name="style">The driving style, or null to derive from the average power.</param>
    /// <param name="updatedAt">The freshness timestamp.</param>
    public static DrivingTipsModel Stale(
        MotorEfficiencyStats stats,
        ThrottleStyle? style = null,
        DateTimeOffset? updatedAt = null)
    {
        ArgumentNullException.ThrowIfNull(stats);
        return new(DrivingTipsState.Stale, stats, style, updatedAt);
    }

    /// <summary>An offline snapshot (no connectivity) carrying the last cached motor stats.</summary>
    /// <param name="stats">The cached motor stats.</param>
    /// <param name="style">The driving style, or null to derive from the average power.</param>
    /// <param name="updatedAt">The freshness timestamp.</param>
    /// <param name="message">An already-localized offline message, or null for the default copy.</param>
    public static DrivingTipsModel Offline(
        MotorEfficiencyStats stats,
        ThrottleStyle? style = null,
        DateTimeOffset? updatedAt = null,
        string? message = null)
    {
        ArgumentNullException.ThrowIfNull(stats);
        return new(DrivingTipsState.Offline, stats, style, updatedAt, ErrorMessage: message);
    }
}

/// <summary>
/// One render-ready recommendation row — the native analogue of a single web tip row (the bordered
/// <c>flex items-start</c> box). Carries the already-localized <see cref="Text"/> and the per-row Narrator
/// <see cref="AutomationName"/>; the leading icon (shield / triangle) is shared across the list and lives on the
/// parent <see cref="DrivingTipsDisplay"/>. Pure data.
/// </summary>
/// <param name="Text">The localized recommendation text (web tip string).</param>
/// <param name="AutomationName">The Narrator name for the row.</param>
public sealed record DrivingTipRow(string Text, string AutomationName);

/// <summary>
/// The fully projected, render-ready view of the driving recommendations — the native analogue of everything the
/// web <c>DrivingTips</c> renders. Holds the active <see cref="State"/>, the localized panel
/// <see cref="Title"/>, the ordered <see cref="Tips"/> rows, the shared leading-icon glyph + tone
/// (<see cref="TipIconGlyph"/> / <see cref="TipIconStatus"/>, the web <c>ShieldCheck</c>-vs-<c>AlertTriangle</c>
/// choice), the freshness chip copy + status (shown only for <see cref="DrivingTipsState.Stale"/> /
/// <see cref="DrivingTipsState.Offline"/>), the loading / error copy and retry label, the freshness timestamp +
/// fetching flag, and the surface <see cref="AutomationName"/>. Pure data so every branch is asserted headlessly.
/// </summary>
/// <param name="State">The resolved render branch.</param>
/// <param name="Title">Localized panel title (web "Driving Style Recommendations").</param>
/// <param name="Tips">The ordered recommendation rows (always at least one).</param>
/// <param name="TipIconGlyph">Segoe Fluent glyph shared by every row (shield when conservative, else triangle).</param>
/// <param name="TipIconStatus">Semantic tone for the row icon (success when conservative, else warning).</param>
/// <param name="ShowFreshnessChip">Whether a stale / offline chip is shown beside the title.</param>
/// <param name="FreshnessChipText">Localized stale / offline chip text.</param>
/// <param name="FreshnessChipStatus">Semantic tone for the freshness chip.</param>
/// <param name="LoadingLabel">Localized loading copy (loading branch).</param>
/// <param name="ErrorTitle">Localized error heading (error branch).</param>
/// <param name="ErrorMessage">Localized error message (error / offline branch).</param>
/// <param name="RetryLabel">Localized retry affordance label.</param>
/// <param name="UpdatedAt">Last successful update timestamp surfaced in the freshness chip.</param>
/// <param name="IsFetching">True while a background refresh is in flight.</param>
/// <param name="AutomationName">The composed Narrator name for the surface.</param>
public sealed record DrivingTipsDisplay(
    DrivingTipsState State,
    string Title,
    IReadOnlyList<DrivingTipRow> Tips,
    string TipIconGlyph,
    StatusKind TipIconStatus,
    bool ShowFreshnessChip,
    string FreshnessChipText,
    StatusKind FreshnessChipStatus,
    string LoadingLabel,
    string ErrorTitle,
    string ErrorMessage,
    string RetryLabel,
    DateTimeOffset? UpdatedAt,
    bool IsFetching,
    string AutomationName);

/// <summary>
/// Pure projection from a <see cref="DrivingTipsModel"/> to its <see cref="DrivingTipsDisplay"/> — the native
/// port of web/src/features/driving/components/driving-dynamics/DrivingTips.tsx. Branch precedence mirrors the
/// web parent's data lifecycle (loading → error → empty → freshness → ready); a fresh snapshot with no stats
/// collapses to the empty branch (the single "drive to collect data" row), while a stale / offline snapshot keeps
/// its cached recommendations under a freshness chip. The recommendation list comes from
/// <see cref="DrivingTipsAdapter.Select"/>; the leading row icon reproduces the web ternary
/// (<c>throttleStyle === 'conservative' ? ShieldCheck : AlertTriangle</c>), deriving the style from the average
/// power when the parent did not supply one (<see cref="ThrottleStyles.FromAveragePower"/>). Every label resolves
/// through the i18n facade using the same keys the web feeds into <c>t(...)</c>. No WinUI types — unit-tested
/// without a UI host.
/// </summary>
public static class DrivingTipsProjection
{
    /// <summary>i18n key for the panel title (web <c>dynamics.recommendations</c>).</summary>
    public const string TitleKey = "dynamics.recommendations";

    /// <summary>English fallback for <see cref="TitleKey"/>.</summary>
    public const string TitleFallback = "Driving Style Recommendations";

    /// <summary>i18n key for the error heading (native-only — the web child has no error branch).</summary>
    public const string ErrorTitleKey = "dynamics.tipsError";

    /// <summary>English fallback for <see cref="ErrorTitleKey"/>.</summary>
    public const string ErrorTitleFallback = "Couldn't load recommendations";

    /// <summary>i18n key for the default error message.</summary>
    public const string ErrorMessageKey = "dynamics.tipsErrorMessage";

    /// <summary>English fallback for <see cref="ErrorMessageKey"/>.</summary>
    public const string ErrorMessageFallback =
        "We couldn't load your driving recommendations. Please try again.";

    /// <summary>Project <paramref name="model"/> into a render-ready display using the i18n facade.</summary>
    /// <param name="model">The render-time data model (the web props plus lifecycle).</param>
    /// <param name="localizer">The i18n facade every label resolves through.</param>
    /// <returns>The render-ready display model.</returns>
    public static DrivingTipsDisplay Project(DrivingTipsModel model, ILocalizer localizer)
    {
        ArgumentNullException.ThrowIfNull(model);
        ArgumentNullException.ThrowIfNull(localizer);

        DrivingTipsState state = SelectState(model);
        string title = localizer.GetString(TitleKey, TitleFallback);

        // web: motorStats drives the tip list; when the snapshot resolved with no stats we show the single
        // no-data row (the empty branch renders the same panel, never a blank box).
        MotorEfficiencyStats? stats = state == DrivingTipsState.Empty ? null : model.Stats;
        List<DrivingTipRow> tips = BuildTips(stats, localizer);

        bool conservative = IsConservative(stats, model.Style);
        string iconGlyph = conservative
            ? DrivingTipsRegistration.ShieldCheckGlyph
            : DrivingTipsRegistration.AlertTriangleGlyph;
        StatusKind iconStatus = conservative ? StatusKind.Success : StatusKind.Warning;

        bool showChip = state is DrivingTipsState.Stale or DrivingTipsState.Offline;
        string chipText = state switch
        {
            DrivingTipsState.Offline => localizer.GetString("common.offline", "Offline"),
            DrivingTipsState.Stale => localizer.GetString("common.stale", "Stale"),
            _ => string.Empty,
        };
        StatusKind chipStatus = state == DrivingTipsState.Offline ? StatusKind.Danger : StatusKind.Warning;

        string loadingLabel = localizer.GetString("common.loading", "Loading");
        string errorTitle = localizer.GetString(ErrorTitleKey, ErrorTitleFallback);
        string errorMessage = string.IsNullOrWhiteSpace(model.ErrorMessage)
            ? localizer.GetString(ErrorMessageKey, ErrorMessageFallback)
            : model.ErrorMessage!;
        string retryLabel = localizer.GetString("common.retry", "Retry");

        string automationName = BuildAutomationName(
            state, title, tips, showChip, chipText, loadingLabel, errorTitle);

        return new DrivingTipsDisplay(
            State: state,
            Title: title,
            Tips: tips,
            TipIconGlyph: iconGlyph,
            TipIconStatus: iconStatus,
            ShowFreshnessChip: showChip,
            FreshnessChipText: chipText,
            FreshnessChipStatus: chipStatus,
            LoadingLabel: loadingLabel,
            ErrorTitle: errorTitle,
            ErrorMessage: errorMessage,
            RetryLabel: retryLabel,
            UpdatedAt: model.UpdatedAt,
            IsFetching: model.IsFetching,
            AutomationName: automationName);
    }

    /// <summary>
    /// Whether the driving style resolves to conservative — the web <c>throttleStyle === 'conservative'</c> test
    /// that picks the shield icon. With no stats the style is null (web parent passes null), so the triangle
    /// shows; otherwise an explicit style wins and a missing one is derived from the average power.
    /// </summary>
    /// <param name="stats">The motor stats, or null.</param>
    /// <param name="style">The parent-supplied style, or null to derive.</param>
    /// <returns>True when the resolved style is conservative.</returns>
    public static bool IsConservative(MotorEfficiencyStats? stats, ThrottleStyle? style)
    {
        if (stats is not { } s)
        {
            // web: throttleStyle = motorStats ? getThrottleStyle(...) : null → null is never 'conservative'.
            return style == ThrottleStyle.Conservative;
        }

        ThrottleStyle resolved = style ?? ThrottleStyles.FromAveragePower(s.AvgPowerKw);
        return resolved == ThrottleStyle.Conservative;
    }

    // Branch precedence from the web parent's data lifecycle. Loading / Error / Empty / Stale / Offline come
    // straight from the parent's classification; a fresh "Ready" snapshot with no stats has only the no-data row
    // and collapses to the empty branch, while a stale / offline snapshot keeps its cached tips.
    private static DrivingTipsState SelectState(DrivingTipsModel model) => model.Status switch
    {
        DrivingTipsState.Loading => DrivingTipsState.Loading,
        DrivingTipsState.Error => DrivingTipsState.Error,
        DrivingTipsState.Empty => DrivingTipsState.Empty,
        DrivingTipsState.Stale => DrivingTipsState.Stale,
        DrivingTipsState.Offline => DrivingTipsState.Offline,
        _ => model.Stats is not null ? DrivingTipsState.Ready : DrivingTipsState.Empty,
    };

    private static List<DrivingTipRow> BuildTips(MotorEfficiencyStats? stats, ILocalizer localizer)
    {
        IReadOnlyList<DrivingTipKind> kinds = DrivingTipsAdapter.Select(stats);
        var rows = new List<DrivingTipRow>(kinds.Count);
        foreach (DrivingTipKind kind in kinds)
        {
            string text = DrivingTipsAdapter.Localize(kind, localizer);
            rows.Add(new DrivingTipRow(text, text));
        }

        return rows;
    }

    private static string BuildAutomationName(
        DrivingTipsState state,
        string title,
        List<DrivingTipRow> tips,
        bool showChip,
        string chipText,
        string loadingLabel,
        string errorTitle)
    {
        switch (state)
        {
            case DrivingTipsState.Loading:
                return string.Create(CultureInfo.CurrentCulture, $"{title}. {loadingLabel}");
            case DrivingTipsState.Error:
                return errorTitle;
            default:
                var parts = new List<string>(tips.Count + 2) { title };
                if (showChip)
                {
                    parts.Add(chipText);
                }

                foreach (DrivingTipRow tip in tips)
                {
                    parts.Add(tip.Text);
                }

                return string.Join(". ", parts);
        }
    }
}

/// <summary>
/// PII-safe diagnostics for the <c>DrivingTips</c> surface (P1/S11 diagnostics contract). Records only the
/// operational <c>view.opened</c> event with the surface slug — never a motor power, temperature or VIN — so a
/// diagnostics line can never leak fleet data. Thread-safe.
/// </summary>
public sealed class DrivingTipsDiagnostics
{
    private readonly Action<string>? _sink;
    private long _viewsOpened;

    /// <summary>Creates the collector over an optional PII-safe diagnostics sink.</summary>
    /// <param name="sink">Optional sink invoked with each diagnostics line.</param>
    public DrivingTipsDiagnostics(Action<string>? sink = null) => _sink = sink;

    /// <summary>Number of times the surface has been opened.</summary>
    public long ViewsOpened => Interlocked.Read(ref _viewsOpened);

    /// <summary>Record that the surface was opened, emitting <c>view.opened slug=DrivingTips</c>.</summary>
    public void RecordViewOpened()
    {
        Interlocked.Increment(ref _viewsOpened);
        _sink?.Invoke($"view.opened slug={DrivingTipsRegistration.Slug}");
    }
}

/// <summary>
/// Canonical metadata for the <c>DrivingTips</c> feature surface — the native mirror of the web component at
/// <c>web/src/features/driving/components/driving-dynamics/DrivingTips.tsx</c>. Holds the diagnostics slug and the
/// Segoe Fluent glyphs that stand in for the web Lucide icons (<c>Lightbulb</c> header, <c>ShieldCheck</c> /
/// <c>AlertTriangle</c> row markers). UI-free so the metadata is asserted in tests.
/// </summary>
public static class DrivingTipsRegistration
{
    /// <summary>Diagnostics surface slug emitted with the <c>view.opened</c> event (P1/S11).</summary>
    public const string Slug = "DrivingTips";

    /// <summary>Segoe Fluent "Lightbulb" glyph for the panel header (web lucide <c>Lightbulb</c>).</summary>
    public const string LightbulbGlyph = "\uEA80";

    /// <summary>Segoe Fluent "Shield" glyph for a conservative-style row (web lucide <c>ShieldCheck</c>).</summary>
    public const string ShieldCheckGlyph = "\uEA18";

    /// <summary>Segoe Fluent "Warning" glyph for a non-conservative-style row (web lucide <c>AlertTriangle</c>).</summary>
    public const string AlertTriangleGlyph = "\uE7BA";
}

File: apps/android/app/src/main/java/com/teslasync/feature-views/DrivingTips/DrivingTipsModel.kt
// Pure, framework-free model + projection for the DrivingTips feature view — the native analogue of
// everything the web component derives before returning JSX
// (web/src/features/driving/components/driving-dynamics/DrivingTips.tsx). No Compose, no Android, no HTTP:
// every declaration here is unit-tested off-device in the :android:testReleaseUnitTest gate, keeping the
// composable a thin render layer.
//
// DrivingTips is a purely presentational, prop-driven surface. The web component takes a `motorStats` object
// (or null) and a `throttleStyle` (or null) — both computed upstream by the Driving Dynamics page from the
// `/motor/...` history — and `useMemo`s an ordered list of coaching tips from them. Its only other dependency
// is `useTranslation` (the P1/S10 i18n catalog). It binds NO data hook of its own, so — exactly as the
// sibling HealthRecommendations / HighlightCard ports reason — there is no loading / error / stale / offline
// lifecycle to model here; that belongs to the owning page, and inventing it would be drift the spec does not
// have. What the web source genuinely varies is the tip list as a function of (`motorStats`, `throttleStyle`),
// and that is exactly what this projection reproduces and the tests exercise.
//
// The two web data branches are the complete state set this surface renders:
//   • an absent `motorStats` (web `if (!motorStats)`) → the single friendly "drive to start collecting" tip,
//     which is the web's own empty handling — a present, readable row, never a blank box; and
//   • a present `motorStats` → one of three power-band tip pairs (high / moderate / efficient), optionally
//     followed by the high-motor-temp thermal tip.
// The leading glyph of every row is chosen by `throttleStyle` (conservative → a reassuring shield-check,
// otherwise an alert triangle), exactly as the web does.
//
// The web `MotorStats` interface (helpers.ts) carries ten fields, but DrivingTips reads only `avgPower` and
// `maxMotorTemp`; following the sibling LiveMotorStatus precedent ("the slice it consumes"), [MotorStats] here
// mirrors just those two. The localized tip text is resolved at the Compose render boundary via the P1/S10
// i18n facade (stringResource) — never stored in this layer — so no English literal lives in the model.
//
// `InvalidPackageDeclaration` is suppressed because this surface's mandated directory
// (com/teslasync/feature-views/DrivingTips — the P3 prompt's allowed-files path) cannot form a valid Kotlin
// package (a hyphen and a PascalCase segment are illegal in a package identifier), so the package
// intentionally diverges from the path — exactly as the sibling HealthRecommendations / LiveMotorStatus
// surfaces do. `MatchingDeclarationName` is suppressed for the co-located supporting declarations.
@file:Suppress("MatchingDeclarationName", "InvalidPackageDeclaration")

package io.teslasync.android.featureviews.drivingtips

import io.teslasync.shared.core.diagnostics.Logger

/** Average drive power (kW) at or above which `getThrottleStyle` classifies the style as aggressive. */
private const val THROTTLE_AGGRESSIVE_MIN_KW: Double = 80.0

/** Average drive power (kW) below which `getThrottleStyle` classifies the style as conservative. */
private const val THROTTLE_CONSERVATIVE_MAX_KW: Double = 20.0

/** Web `motorStats.avgPower > 80` — above this the surface shows the high-power "ease off" coaching pair. */
private const val TIP_HIGH_POWER_KW: Double = 80.0

/** Web `motorStats.avgPower > 20` — above this (but not high) the surface shows the moderate-power pair. */
private const val TIP_MODERATE_POWER_KW: Double = 20.0

/** Web `motorStats.maxMotorTemp > 120` — above this the thermal-caution tip is appended. */
private const val TIP_HIGH_MOTOR_TEMP_C: Double = 120.0

/**
 * Throttle-input classification — the native analogue of the web `ThrottleStyle` union
 * (`'conservative' | 'moderate' | 'aggressive'`, from `driving-dynamics/helpers.ts`). The web value is
 * computed upstream by `getThrottleStyle(avgPower)` and threaded into DrivingTips as the `throttleStyle` prop;
 * this surface reads it only to pick each tip row's leading glyph (conservative → shield-check, otherwise
 * alert triangle).
 */
enum class ThrottleStyle {
    Conservative,
    Moderate,
    Aggressive,
    ;

    companion object {
        /**
         * Maps a raw `throttleStyle` union string to its [ThrottleStyle]. The web keys are exact lowercase; an
         * absent (`null`) or unrecognised value folds to `null`, mirroring the web prop's `ThrottleStyle | null`
         * type so an upstream gap renders the neutral (alert-triangle) glyph rather than a fabricated style.
         */
        fun fromRaw(value: String?): ThrottleStyle? =
            when (value) {
                "conservative" -> Conservative
                "moderate" -> Moderate
                "aggressive" -> Aggressive
                else -> null
            }

        /**
         * Reproduces the web `getThrottleStyle(avgPower)` derivation (helpers.ts): `< 20` → [Conservative],
         * `< 80` → [Moderate], else [Aggressive]. Exposed so the owning Driving Dynamics surface can compute the
         * `throttleStyle` prop from the same `avgPower` it feeds [MotorStats], keeping the two in lock-step.
         */
        fun fromAvgPower(avgPower: Double): ThrottleStyle =
            when {
                avgPower < THROTTLE_CONSERVATIVE_MAX_KW -> Conservative
                avgPower < THROTTLE_AGGRESSIVE_MIN_KW -> Moderate
                else -> Aggressive
            }
    }
}

/**
 * The slice of the web `MotorStats` (helpers.ts) that DrivingTips actually reads — the average drive power and
 * the peak motor temperature, both already in their display units (kW and °C, derived upstream exactly as the
 * web renders them). Modelled as the consumed slice rather than the full ten-field interface, following the
 * sibling LiveMotorStatus precedent, because these two values fully determine the tip list.
 *
 * @property avgPower average motor power in kW (web `motorStats.avgPower`) — selects the power-band tip pair.
 * @property maxMotorTemp peak motor temperature in °C (web `motorStats.maxMotorTemp`) — gates the thermal tip.
 */
data class MotorStats(
    val avgPower: Double,
    val maxMotorTemp: Double,
)

/**
 * A single coaching tip DrivingTips can emit — its stable identity ([listKey]) and its i18n key ([i18nKey]).
 * Mirrors the eight `t('dynamics.tip*', …)` strings the web component can push. The localized text is resolved
 * at the Compose boundary, never stored here, so the model holds no English literal. The constant order is the
 * canonical web emission order; [DrivingTipsProjection] selects and orders the subset shown for a given input.
 *
 * @property listKey a stable, locale-independent id used as the Compose list key and in tests (the native
 *   analogue of the web `key={i}` index, but content-stable).
 * @property i18nKey the P1/S10 catalog key whose value the composable resolves for the tip text
 *   (e.g. `dynamics.tipNoData`).
 */
enum class DrivingTip(
    val listKey: String,
    val i18nKey: String,
) {
    NoData("no-data", "dynamics.tipNoData"),
    EaseAccel("ease-accel", "dynamics.tipEaseAccel"),
    BrakeEarly("brake-early", "dynamics.tipBrakeEarly"),
    SmoothThrottle("smooth-throttle", "dynamics.tipSmoothThrottle"),
    Coast("coast", "dynamics.tipCoast"),
    Great("great", "dynamics.tipGreat"),
    Keep("keep", "dynamics.tipKeep"),
    Thermal("thermal", "dynamics.tipThermal"),
}

/**
 * Pure projection from the surface's inputs to its ordered tip list — a 1:1 port of the web component's
 * `useMemo` body. Reproduces the exact branch order:
 *
 *   1. no `motorStats` → just the "drive to start collecting" tip (the web empty handling), then stop;
 *   2. `avgPower > 80` → ease-accel + brake-early; else `avgPower > 20` → smooth-throttle + coast;
 *      else → great + keep;
 *   3. `maxMotorTemp > 120` → append the thermal-caution tip.
 *
 * so the same input always yields the same list, in the same order, as the web. Stateless and Compose-free.
 */
object DrivingTipsProjection {
    /**
     * Build the ordered tip list for [motorStats], mirroring the web `useMemo`. Returns exactly one tip when
     * [motorStats] is `null` (the friendly no-data row), otherwise two power-band tips plus an optional third
     * thermal tip — so the result is never empty and the panel is never a blank box.
     */
    fun tipsFor(motorStats: MotorStats?): List<DrivingTip> {
        if (motorStats == null) return listOf(DrivingTip.NoData)
        return buildList {
            when {
                motorStats.avgPower > TIP_HIGH_POWER_KW -> {
                    add(DrivingTip.EaseAccel)
                    add(DrivingTip.BrakeEarly)
                }
                motorStats.avgPower > TIP_MODERATE_POWER_KW -> {
                    add(DrivingTip.SmoothThrottle)
                    add(DrivingTip.Coast)
                }
                else -> {
                    add(DrivingTip.Great)
                    add(DrivingTip.Keep)
                }
            }
            if (motorStats.maxMotorTemp > TIP_HIGH_MOTOR_TEMP_C) {
                add(DrivingTip.Thermal)
            }
        }
    }
}

/**
 * The one PII-safe diagnostic this surface emits (P1/S11). Carries only the surface [SLUG] — never the motor
 * stats, the throttle style, or any tip text — so a diagnostics line can never leak vehicle state.
 */
object DrivingTipsDiagnostics {
    /** Diagnostics surface slug emitted with the `view.opened` event. */
    const val SLUG: String = "DrivingTips"

    private const val VIEW_OPENED = "view.opened"
    private const val SURFACE_KEY = "surface"

    /** Emits the `view.opened` diagnostic for this surface. Call from the composable's first-composition effect. */
    fun recordViewOpened(logger: Logger) {
        logger.info(VIEW_OPENED, mapOf(SURFACE_KEY to SLUG))
    }
}
